### PR TITLE
Add backflow treatment to IPCS_ABCN sovler

### PR DESCRIPTION
**Pull request by hkjeldsberg**
- Added option to enable backflow treatment method (Outlet stabilization from Moghadam et al. Comput Mech (2011) 48:277–291) to IPCS_ABCN solver
-  Added option to enable Neumann boundary condition to IPCS_ABCN solver 

P.S. Unsure if we're using Neumann condition (open boundary) or p=0 at outlet for stenosis simulation.

### CODE REVIEW

**Comment by hkjeldsberg** [comment]
@johannesring Usage: 

```
NS_parameters.update(
        # Backflow
        backflow_facets=[3]  # Apply backflow stabilization to outlet with id=3 
)
```

File: oasis/problems/NSfracStep/__init__.py [hunk 1]
```diff
@@ -60,6 +60,13 @@
         rho=1085        # kg/m^3
         ),
 
+    # Neumann boundary condition on the outlets
+    neumann_facets=[],  # List of facets value(s) in boundaries to apply boundary condition
+
+    # Back flow stabilization, turned on if back_flow_facets is != [] and boundaries is not None
+    backflow_facets=[],  # List of facet value(s) in MeshFunction (boundaries) to apply back flow stabilization
+    backflow_beta=0.2,  # Standard value from Moghadam et al. Comput Mech (2011) 48:277–291
+
```
Comment (johannesring) [inline oasis/problems/NSfracStep/__init__.py:69]: @hkjeldsberg - Is there a missing default value for `boundary`? I get this error:
```
Traceback (most recent call last):
  File "/cluster/home/johannr/src/Oasis_backflow/local/bin/oasis", line 11, in <module>
    load_entry_point('oasis==2018.1', 'console_scripts', 'oasis')()
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/run_oasis.py", line 11, in main
    from oasis import NSfracStep
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/NSfracStep.py", line 169, in <module>
    vars().update(setup(**vars()))
TypeError: setup() missing 1 required positional argument: 'boundary'
```

Comment (hkjeldsberg) [reply oasis/problems/NSfracStep/__init__.py:69]: Good point! Yes, I think the `boundary` needs to be part of the namespace. A possibility is to add the following to the problem file
```
def pre_boundary_condition(mesh, **NS_namespace):
    boundaries = MeshFunction("size_t", mesh, 2, mesh.domains())
    boundaries.set_values(boundaries.array() + 1)

    return dict(boundary=boundaries)
```

with `pre_boundary_condition()` being called before boundary conditions are set (by default).

Comment (johannesring) [reply oasis/problems/NSfracStep/__init__.py:69]: Do you mean to add something like the following?
```
diff --git a/oasis/NSfracStep.py b/oasis/NSfracStep.py
index cd6c9ed..17180ea 100755
--- a/oasis/NSfracStep.py
+++ b/oasis/NSfracStep.py
@@ -131,6 +131,9 @@ for ci in scalar_components:
 
 print_solve_info = use_krylov_solvers and krylov_solvers['monitor_convergence']
 
+# Pre boundary condition setup
+vars().update(pre_boundary_condition(**vars()))
+
 # Boundary conditions
 bcs = create_bcs(**vars())
 
diff --git a/oasis/problems/__init__.py b/oasis/problems/__init__.py
index 42d06f2..affb2ce 100644
--- a/oasis/problems/__init__.py
+++ b/oasis/problems/__init__.py
@@ -166,6 +166,11 @@ def initialize(**NS_namespace):
     pass
 
 
+def pre_boundary_condition(**NS_namespace):
+    """Called prior to create_bcs. Must return a dictionary."""
+    return dict(boundary=None)
+
+
 def create_bcs(sys_comp, **NS_namespace):
     """Return dictionary of Dirichlet boundary conditions."""
     return dict((ui, []) for ui in sys_comp)
```
I tried that, but then I got this error message:
```
Start simulations
Traceback (most recent call last):
  File "/cluster/home/johannr/src/Oasis_backflow/local/bin/oasis", line 11, in <module>
    load_entry_point('oasis==2018.1', 'console_scripts', 'oasis')()
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/run_oasis.py", line 11, in main
    from oasis import NSfracStep
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/NSfracStep.py", line 196, in <module>
    assemble_first_inner_iter(**vars())
TypeError: assemble_first_inner_iter() missing 1 required positional argument: 'NT'
```

Comment (hkjeldsberg) [reply oasis/problems/NSfracStep/__init__.py:69]: Yes, I believe the hook already exists, but it needs to be specifically defined in the problem file (`stenosis_backflow.py`).
I think my code snippet should work for that problem, specifically.

Odd that it's complaining about missing 'NT'. Which solver are you using?

Comment (johannesring) [reply oasis/problems/NSfracStep/__init__.py:69]: Yes, I added the code in your [first comment](https://github.com/johannesring/Oasis/pull/1/files#r569370006) to the problem file (`stenosis_backflow.py`), but I could not find the hook (so I added it). Can you point me to where it is?

Sorry, my fault about the missing `NT`. Now I got this error message:
```
Adding Backflow stabilization for boundary with id(s)=[3]
Invalid cell <module 'dolfin.mesh' from '/cluster/shared/fenics/nobackup/install/fenics-2019.1.0-2020-02-19/lib/python3.6/site-packages/dolfin/mesh/__init__.py'>.
Traceback (most recent call last):
  File "/cluster/home/johannr/src/Oasis_backflow/local/bin/oasis", line 11, in <module>
    load_entry_point('oasis==2018.1', 'console_scripts', 'oasis')()
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/run_oasis.py", line 11, in main
    from oasis import NSfracStep
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/NSfracStep.py", line 172, in <module>
    vars().update(setup(**vars()))
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/solvers/NSfracStep/IPCS_ABCN.py", line 101, in setup
    ds = Measure("ds", domain=mesh, subdomain_data=boundary)
  File "/cluster/shared/fenics/nobackup/install/fenics-2019.1.0-2020-02-19/lib/python3.6/site-packages/ufl/measure.py", line 158, in __init__
    self._domain = None if domain is None else as_domain(domain)
  File "/cluster/shared/fenics/nobackup/install/fenics-2019.1.0-2020-02-19/lib/python3.6/site-packages/ufl/domain.py", line 296, in as_domain
    cell = as_cell(domain)
  File "/cluster/shared/fenics/nobackup/install/fenics-2019.1.0-2020-02-19/lib/python3.6/site-packages/ufl/cell.py", line 329, in as_cell
    error("Invalid cell %s." % cell)
  File "/cluster/shared/fenics/nobackup/install/fenics-2019.1.0-2020-02-19/lib/python3.6/site-packages/ufl/log.py", line 172, in error
    raise self._exception_type(self._format_raw(*message))
ufl.log.UFLException: Invalid cell <module 'dolfin.mesh' from '/cluster/shared/fenics/nobackup/install/fenics-2019.1.0-2020-02-19/lib/python3.6/site-packages/dolfin/mesh/__init__.py'>.
```
Do you know why?

Comment (hkjeldsberg) [reply oasis/problems/NSfracStep/__init__.py:69]: Oh, my mistake! The 'pre_boundary_condition()'- hook is not there as default, but your implementation looks correct. 

I have a similar hook for OasisMove: 

```
def pre_boundary_condition(**NS_namespace):
    """Called after defining the function spaces and functions,
       but before the boundary
       conditions.
    """
    return {}
```

I noticed that `mesh` is missing as an input parameter to the `setup()`-method inside `IPCS_ABCN.py`. This should fix it.  Right now I think the solver is using the 'mesh()'-function at line 101.

Comment (johannesring) [reply oasis/problems/NSfracStep/__init__.py:69]: Adding `mesh` fixed the problem, but now I got the problem with missing `NT` again:
```

Start simulations
Traceback (most recent call last):
  File "/cluster/home/johannr/src/Oasis_backflow/local/bin/oasis", line 11, in <module>
    load_entry_point('oasis==2018.1', 'console_scripts', 'oasis')()
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/run_oasis.py", line 11, in main
    from oasis import NSfracStep
  File "/cluster/home/johannr/src/Oasis_backflow/local/lib/python3.6/site-packages/oasis/NSfracStep.py", line 197, in <module>
    assemble_first_inner_iter(**vars())
TypeError: assemble_first_inner_iter() missing 1 required positional argument: 'NT'
```

Comment (hkjeldsberg) [reply oasis/problems/NSfracStep/__init__.py:69]: Ah, the update must've overwritten something in `IPCS_ABCN.py` (line 110): 

```
d.update(u_ab=u_ab, a_conv=a_conv, a_scalar=a_scalar, LT=LT, KT=KT, K2=K2)
```
should probably be changed to
```
d.update(u_ab=u_ab, a_conv=a_conv, a_scalar=a_scalar, LT=LT, KT=KT, K2=K2, NT=NT)
```

**Pull request merged**